src/main/java/SpringWeatherForecast/WeatherController.java

```diff
@@ -0,0 +1,38 @@
+package SpringWeatherForecast;
+
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+public class WeatherController {
+    private final WeatherRepository repository;
+
+    DataSource source = new DataSource();
+    WeatherForecaster forecaster = new WeatherForecaster(source);
```

Comment:
These 3 dependencies: data source, forecaster and repository are too much for the controller. The idea is to encapsulate them in a kind of WeatherService, which in turn works with either repository or an external weather service.